File: controllers/PageController.php
<?php
class PageController extends Controller
{
    public function init()
    {
        Session::init();
    }

    public function approot_js()
    {
        $data = array(
            'root' => Config::$config['root_path'],
            'script_name' => Config::$config['script_name']
        );
        return $this->response->renderView('pages.approot_js', $data, 'text/javascript');
    }

    public function index()
    {
        if (Session::loggedIn()) {
            return $this->response->redirect(vh_link('page/dashboard'));
        }

        $data = array(
            'title' => 'Home - MadToDo',
            'isHome' => true,
            'headerScripts' => array(
                'js/madtodo.js',
                'js/xhr.js',
                'js/user.js',
            ),
            'footerScripts' => array(
                'js/home.js',
            )
        );

        return $this->response->renderView('pages.home', $data);
    }

    public function home()
    {
        return $this->response->redirect(vh_link(''));
    }

    public function dashboard()
    {
        if (!Session::loggedIn()) {
            return $this->response->redirect(vh_link(''));
        }

        App::loadModel('User');
        $user = User::getOneByUsername(Session::get('username'));

        $data = array(
            'title' => 'Dashboard - MadToDo',
            'isDashboard' => true,
            'headerScripts' => array(
                'js/madtodo.js',
                'js/xhr.js',
                'js/user.js',
                'js/task.js',
            ),
            'footerScripts' => array(
                'js/dashboard.js',
                'js/dialog.js',
            ),
            'user' => $user->toArray(),
        );

        return $this->response->renderView('pages.dashboard', $data);
    }

    public function profile()
    {
        if (!Session::loggedIn()) {
            return $this->response->redirect(vh_link(''));
        }

        App::loadModel('User');
        $user = User::getOneByUsername(Session::get('username'));

        $data = array(
            'title' => 'Profile - MadToDo',
            'isProfile' => true,
            'headerScripts' => array(
                'js/madtodo.js',
                'js/xhr.js',
                'js/user.js',
                'js/task.js',
            ),
            'footerScripts' => array(
                'js/profile.js',
                'js/dialog.js',
            ),
            'user' => $user->toArray(),
        );

        return $this->response->renderView('pages.profile', $data);
    }

    public function search($filter="", $keyword="")
    {
        if (!Session::loggedIn()) {
            return $this->response->redirect(vh_link(''));
        }

        App::loadModel('User');
        App::loadModel('Category');
        App::loadModel('Task');
        App::loadModel('Tag');
        App::loadModel('Comment');

        $user = User::getOneByUsername(Session::get('username'));

        $data = array(
            'title' => 'Profile - MadToDo',
            'isSearch' => true,
            'headerScripts' => array(
                'js/madtodo.js',
                'js/xhr.js',
                'js/user.js',
                'js/task.js',
                'js/search.js',
            ),
            'footerScripts' => array(
                'js/search.js',
            ),
            'user' => $user->toArray(),
            'filter' => $filter,
            'keyword' => $keyword,
        );

        if ($filter == 'all') {
            $users = User::getAll(array(
                'where' => array(
                    array('username', 'LIKE', '%' . $keyword . '%'),
                ),
            ), false);
            $data['users'] = $users;

            $categories = Category::getAll(array(
                'where' => array(
                    array('name', 'LIKE', '%' . $keyword . '%'),
                    'AND',
                    array('user_id', '=', $user->get_id()),
                ),
            ), false);
            $data['categories'] =  $categories;

            $tasks = Task::getAll(array(
                'where' => array(
                    array('name', 'LIKE', '%' . $keyword . '%'),
                ),
            ));

            $data['tasks'] = $tasks;
        }
        else if ($filter == 'username') {
            $users = User::getAll(array(
                'where' => array(
                    array('username', 'LIKE', '%' . $keyword . '%'),
                ),
            ), false);
            $data['users'] = $users;
        }
        else if ($filter == 'category') {
            $categories = Category::getAll(array(
                'where' => array(
                    array('name', 'LIKE', '%' . $keyword . '%'),
                    'AND',
                    array('user_id', '=', $user->get_id()),
                ),
            ), false);
            $data['categories'] =  $categories;
        }
        else if ($filter == 'task') {
            $tasks = Task::getAll(array(
                'where' => array(
                    array('name', 'LIKE', '%' . $keyword . '%'),
                ),
            ));

            $data['tasks'] = $tasks;
        }
        else {
            return $this->response->nullJson();
        }

        return $this->response->renderView('pages.search', $data);
    }

    // Private functions
    public function _complete(&$task) {
            $task->user = $task->get_user()->get_name();
            $assignee = $task->get_assignee();
            if ($assignee != null) {
                $task->assignee = $assignee->get_name();
            }
            $task->category = $task->get_category()->get_name();
            $task->tags = $task->get_tags(true);
            $task->comments = $task->get_comments(true);
    }

    /* temporary: */
    public function install()
    {
        return $this->response->redirect(vh_slink('install.html'));
    }

    public function test()
    {
        return $this->response->redirect(vh_link('hello/view/'
            . urlencode('The quick<br/>brown fox<br/>jumps over<br/>the lazy<br/>dog')));
    }

    public function logout()
    {
        Session::logout();
        return $this->response->redirect(vh_link(''));
    }
}

File: src/madtodo/controllers/PageController.java
package madtodo.controllers;

import madtodo.MadController;
import madtodo.models.Category;
import madtodo.models.Task;
import madtodo.models.User;

public class PageController extends MadController {
    /**
     * url: /
     */
    @Override
    public void index() {
        if (session.isLoggedIn()) {
            sendRedirect("/page/dashboard");
        } else {
            this
            .setAttr("pageTitle", "Home - MadToDo")
            .setAttr("isHome", true)
            .setAttr("headerScripts", new String[] {
                    "/static/js/madtodo.js",
                    "/static/js/xhr.js",
                    "/static/js/user.js"
            })
            .setAttr("footerScripts", new String[] {
                    "/static/js/home.js"
            });

            renderJSPView("/pages/c-index.jsp");
        }
    }

    /**
     * url: /page/dashboard
     */
    public void dashboard() {
        if (!session.isLoggedIn()) {
            sendRedirect("/");
        } else {
            final User user = User.findById(session.getInt("userId"));

            this
            .setAttr("pageTitle", "Dashboard - MadToDo")
            .setAttr("isDashboard", true)
            .setAttr("headerScripts", new String[] {
                    "/static/js/madtodo.js",
                    "/static/js/xhr.js",
                    "/static/js/user.js",
                    "/static/js/task.js"
            })
            .setAttr("footerScripts", new String[] {
                    "/static/js/dashboard.js",
                    "/static/js/dialog.js"
            })
            .setAttr("user", user)
            .setAttr("isUserSet", true);

            renderJSPView("/pages/c-dashboard.jsp");
        }
    }

    /**
     * url: /page/profile
     */
    public void profile() {
        if (!session.isLoggedIn()) {
            sendRedirect("/");
        } else {
            final User user = User.findById(session.getInt("userId"));

            this
            .setAttr("pageTitle", "Profile - MadToDo")
            .setAttr("isProfile", true)
            .setAttr("headerScripts", new String[] {
                    "/static/js/madtodo.js",
                    "/static/js/xhr.js",
                    "/static/js/user.js",
                    "/static/js/task.js"
            })
            .setAttr("footerScripts", new String[] {
                    "/static/js/profile.js",
                    "/static/js/dialog.js"
            })
            .setAttr("user", user)
            .setAttr("isUserSet", true);

            renderJSPView("/pages/c-profile.jsp");
        }
    }

    /**
     * url: /page/search/:filter/[:keyword]
     */
    public void search() {
        final String filter = getParam(0, null);
        final String keyword = getParam(1, "");

        if (!session.isLoggedIn() || filter == null) {
            sendRedirect("/");
        } else {
            int userId = session.getInt("userId");

            if (filter.equals("all")) {
                User[] users = User.searchByUsername(keyword);
                Category[] cats = Category.searchByName(keyword, userId);
                Task[] tasks = Task.searchByName(keyword);

                setAttr("users", users);
                setAttr("cats", cats);
                setAttr("tasks", tasks);
            }
            else if (filter.equals("username")) {
                User[] users = User.searchByUsername(keyword);
                setAttr("users", users);
            }
            else if (filter.equals("category")) {
                Category[] cats = Category.searchByName(keyword, userId);
                setAttr("cats", cats);
            }
            else if (filter.equals("task")) {
                Task[] tasks = Task.searchByName(keyword);
                setAttr("tasks", tasks);
            }
            else {
                print404();
                return;
            }

            User user = User.findById(userId);

            this
            .setAttr("pageTitle", "Search - MadToDo")
            .setAttr("isSearch", true)
            .setAttr("headerScripts", new String[] {
                    "/static/js/madtodo.js",
                    "/static/js/xhr.js",
                    "/static/js/user.js",
                    "/static/js/task.js",
            })
            .setAttr("footerScripts", new String[] {
                    "/static/js/search.js",
            })
            .setAttr("user", user)
            .setAttr("isUserSet", true)
            .setAttr("filter", filter)
            .setAttr("keyword", keyword);

            renderJSPView("/pages/c-search.jsp");
        }
    }

    /**
     * url: /page/logout
     */
    public void logout() {
        if (session.isLoggedIn()) {
            session.logout();
        }

        sendRedirect("/");
    }

    // For testing purpose:

    /**
     * url: /page/jeyesp
     */
    public void jeyesp() {
        String[] list = new String[] {
                "Foo", "Bar", "Qux" };

        setAttr("title", "Jeyesp Test");
        setAttr("list", list);
        setAttr("params", getParams());

        renderJSPView("/pages/jeyesp.jsp");
    }
}
